packages/flutter_tools/lib/src/android/android_studio.dart

```diff
@@ -475,4 +476,14 @@ class AndroidStudio implements Comparable<AndroidStudio> {
 
   @override
   String toString() => 'Android Studio ($version)';
+
+  @override
+  // ignore: avoid_equals_and_hash_code_on_mutable_classes
+  int get hashCode => directory.hashCode;
+
+  @override
+  // ignore: avoid_equals_and_hash_code_on_mutable_classes
+  bool operator ==(Object other) {
+    return other is Object && hashCode == other.hashCode;
```

Comment:
`==` should check the other object is also an `AndroidStudio` object, not `Object`.  `==` needs to actually check equality, not hashCodes.  `==` needs to be reliable even with hash collisions.
```suggestion
    return other is AndroidStudio && directory == other.directory;
```


> // ignore: avoid_equals_and_hash_code_on_mutable_classes

I believe this class could be made immutable if the `_validationMessages` population were hoisted into the factory method, then `_validationMessages`, and `_isValid`  passed in as an immutable `List`.  However, instead of doing all that, I think the right thing is to remove your `hashCode` and `==` changes, and instead fix the existing `FileSystemEntity` checks.